### PR TITLE
ODC-6660: Render topology differently based on zoom level

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
**Fixes**: 
Implements [ODC-6660](https://issues.redhat.com/browse/ODC-6660)

**Analysis / Root cause**: 
As a user, I want the topology view to be less cluttered as I doom out showing only information that I can discern and still be able to get a feel for the status of my project.

**Solution Description**: 

- When zoomed to medium scale, all labels are hidden. Label are shown when hovering over the node.
- When zoomed to 30% scale, all labels, decorators, pod rings & icons will be hidden. Node shape remains the same, and background is either white, yellow or red. Background color is determined based on the worst status of pods, build, or alerts. When hovering over the node, it is scaled up to 1:1 scale so the entire node, label, and decorators are shown.

**Screen shots / Gifs for design review**: 
console-zoom-details

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [x] Safari
- [x] Edge

### CODE REVIEW

**Comment by serenamarie125** [comment]
@jeff-phillips-18 how does aggregate status work?  we should be showing the most severe if there are multiple contributors

**Comment by jeff-phillips-18** [comment]
@serenamarie125 Updated:
console-zoom-details

File: frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx [hunk 1]
```diff
@@ -13,18 +15,134 @@ import {
...
+    'Ready',
+    'Up to date',
+    'Provisioned as node',
+    'Preferred',
+    'Connected',
+    'Info',
+    'Unknown',
+    'PipelineNotStarted',
+  ],
+};
```
Comment (jeff-phillips-18) [inline frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx:100]: @karthikjeeyar Could you verify these build status categories? They are based on https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/app/components/status/Status.tsx#L32

**Review by karthikjeeyar — COMMENTED** [review]
This looks great! ❤️ 

Here are few observations while testing the PR:

1. Build Failure and Pipeline Failure should be marked as red. (but it is mapped to default white color).

build-and-pipeline-failure

---

2. Selected node blue fill color is missing for shapes other than circle.

Base node - Circle (has blue fill color):
image

Other Shapes - Rhombus: (blue fill is missing)
image

File: frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx [hunk 2]
```diff
@@ -13,18 +15,134 @@ import {
...
+
+export const getAggregateStatus = (
+  donutStatus: PodRCData,
+  alertSeverity: AlertSeverity,
+  buildStatus: string,
+): NodeStatus => {
+  if (!donutStatus?.pods?.length) {
+    return NodeStatus.default;
+  }
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx:122]: I think, this early return based on pods length is skipping the status for build failure or pipeline failure (Pods length will be always zero in this case).

File: frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx [hunk 3]
```diff
@@ -13,18 +15,134 @@ import {
...
+    'InstallReady',
+    'Replacing',
+    'Running',
+    'Updating',
+    'Upgrading',
+    'Cancelled',
+    'Deleting',
+    'Expired',
+    'Not Ready',
+    'Terminating',
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx:82]: This block exists in the warning list, so shall we remove it from here?

File: frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx [hunk 4]
```diff
@@ -13,18 +15,134 @@ import {
...
+    return NodeStatus.default;
+  }
+  const worstPodStatus = donutStatus.pods.reduce((acc, pod) => {
+    return Math.max(acc, getNodePodStatus(getPodStatus(pod)));
+  }, POD_STATUS_NORMAL);
+
+  if (
+    worstPodStatus === POD_STATUS_DANGER ||
+    alertSeverity === AlertSeverity.Critical ||
+    buildStatusSeverities.danger.includes[buildStatus]
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx:130]:
```suggestion
    buildStatusSeverities.danger.includes(buildStatus)
```

File: frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx [hunk 5]
```diff
@@ -13,18 +15,134 @@ import {
...
+  ) {
+    return NodeStatus.danger;
+  }
+  if (
+    worstPodStatus === POD_STATUS_WARNING ||
+    alertSeverity === AlertSeverity.Warning ||
+    buildStatusSeverities.warning.includes[buildStatus]
```
Comment (karthikjeeyar) [inline frontend/packages/topology/src/components/graph-view/components/nodes/WorkloadNode.tsx:137]:
```suggestion
    buildStatusSeverities.warning.includes(buildStatus)
```

**Comment by jeff-phillips-18** [comment]
Thanks @karthikjeeyar! Updated.

**Comment by karthikjeeyar** [comment]
Verified on clusterbot with large number of workloads. It works as expected.

image

on node Hover, the connector is placed inside the node. As discussed this issue will be fixed on the next PF update. 

image (1)

/lgtm

**Comment by invincibleJai** [comment]
Propagate from the Epic
/label docs-approved
/label px-approved

**Comment by jerolimov** [comment]
Hey @jeff-phillips-18 looks fine also with 100 nodes, but I get sometimes this strange effect:

Peek 2022-06-16 01-54

(I just moved the mouse, never clicked or so.)

Do you can / want take a look into this? Tested with Chrome 102 on Linux, on a cluster bot instance.

**Comment by jeff-phillips-18** [comment]
Updated to fix create connector when scaled up at medium scale:

med-scale-create-connectors

**Comment by jeff-phillips-18** [comment]
> Hey @jeff-phillips-18 looks fine also with 100 nodes, but I get sometimes this strange effect:
> 
> (I just moved the mouse, never clicked or so.)
> 
> Do you can / want take a look into this? Tested with Chrome 102 on Linux, on a cluster bot instance.

I'm not seeing this on my mac with Chrome 102. I have 600 nodes and the hovering works fine every time. Unsure what the difference is.

**Comment by jerolimov** [comment]
Hey @jeff-phillips-18, looks like the cypress tests still don't work as expected.

Btw, I saw that the decorator flickers on hover. Maybe because there is a new node rendered? Should I file this as a bug, do you want to take a look after this PR here?

Peek 2022-06-20 09-25

File: frontend/packages/topology/integration-tests/support/pages/topology/topology-page.ts [hunk 6]
```diff
@@ -268,11 +268,10 @@ export const topologyPage = {
...
       .contains(serviceName);
   },
-  getKnativeRevision: (serviceName: string) => {
+  getKnativeRevision: (revisionName: string) => {
     return cy
-      .get('[data-type="knative-service"]')
-      .find('[data-type="knative-revision"] g.odc-workload-node')
-      .contains(serviceName);
+      .get('[data-type="knative-revision"]')
+      .find(`[data-test-id="${revisionName}"] g.odc-workload-node`);
```
Comment (invincibleJai) [inline frontend/packages/topology/integration-tests/support/pages/topology/topology-page.ts:274]: e2e is failing here looks like

Comment (jeff-phillips-18) [reply frontend/packages/topology/integration-tests/support/pages/topology/topology-page.ts]: Thanks @invincibleJai, I can't get the local tests to run so I've been struggling.

Comment (invincibleJai) [reply frontend/packages/topology/integration-tests/support/pages/topology/topology-page.ts]: Thanks Jeff, waiting for it to get in 👍

**Pull request merged**